xml_to_json.py

```diff
@@ -1,3 +1,96 @@
+
+    return xmltodict.parse(data)
+
+
+def convert_xml_file_to_json(xml_file_path: str, delete_xmls: bool):
+    values_dict = xml_path_to_dict(xml_file_path)
+    json_string = json.dumps(values_dict, sort_keys=True, indent=4)
+
+    json_file_path = xml_file_path.split('.')[0] + '.json'
+
+    with open(json_file_path, 'w+') as json_file:
+        json_file.write(json_string)
+
+    if delete_xmls:
+        os.remove(xml_file_path)
+
+
+def is_filename_hidden(filename: str) -> bool:
+    return filename[0] == '.' and len(filename) > 1
+
+
+def is_file_within_legal_depth(root_dir: str, file: str, recursion_depth: Optional[int]) -> bool:
+    return recursion_depth is None or len(file.split('\\')) - len(root_dir.split('\\')) <= recursion_depth
+
+
+def all_xmls(root_directory_path: str, recursion_depth: Optional[int]) -> str:
+    for root, dirs, files in os.walk(root_directory_path):
+        # remove from our search tree all hidden directories and those that exceed the recursion depth
+        sub_dirs_to_remove: List[str] = [sub_dir for sub_dir in dirs if
+                                         is_filename_hidden(sub_dir.split('\\')[-1]) or not is_file_within_legal_depth(
+                                             root_directory_path, r"{}\{}".format(root, sub_dir), recursion_depth)]
+
+        for sub_dir_to_remove in sub_dirs_to_remove:
+            dirs.remove(sub_dir_to_remove)
+
+        for xml_file_name in filter(lambda file: file.endswith('.xml'), files):
+            yield r"{}\{}".format(root, xml_file_name)
+
+
+@click.command()
+@click.argument('path', type=click.Path(exists=True))
+@click.option('-r', '--recursion-depth', 'recursion_depth', type=click.INT, default=None,
+              help="""The depth of the recursion, unlimited by default,
+              use a value of 0 for no recursion (i.e only in the PATH directory)""")
+@click.option('-d', '--delete-xmls', 'delete_xmls', is_flag=True,
+              help="Whether or not we want to delete the xml files after the conversion")
+@click.option('-w', '--num-workers', 'num_workers', type=click.IntRange(1, 20, clamp=True), default=1,
+              help="The number of workers that will convert files. Values not in [1,20] will be clamped.")
+def convert_xml_files_to_json(path: str, recursion_depth: Optional[int], delete_xmls: bool, num_workers: int):
+    """
+    converts XML files to JSON files
+
+    PATH the path to either the file you want to convert or the directory where you want to convert all files
+    """
+
+    if os.path.isdir(path):
+        to_convert = list(all_xmls(path, recursion_depth))
+
+        if len(to_convert) == 0:
+            raise Exception("ERROR: No XML files found!")
+
+        pool = Pool(num_workers)
+        pbar = tqdm(desc="XML to JSON Conversion Progress", total=len(to_convert))
+        for xml_file in to_convert:
+            pool.apply_async(convert_xml_file_to_json, args=(xml_file, delete_xmls), callback=lambda _: pbar.update())
```

Comment:
I’m not familiar with pool but from a quick look you turn it on the same file?!

Reply:
nope, for each file I'm adding a task to the pool for converting it